### DISCUSSION TITLE
Preventing update on stale data

### DISCUSSION THREAD

**Discussion by einarpersson** [pocketbase/pocketbase#1884] [Q&A] [answered]
Imagine a collection `organizations` which can be edited by users `Alice` and `Bob`. Can I add a rule somehow that prevents one user from trying to update a entity in the collection based on stale data? (i.e. Alice has already changed a field, and then Bob trys to edit the same field without knowing anything about Alice's changes)

I understand that I can use the realtime/subscription functionality to try to keep clients always up-to-date, but I wonder if it is possible to add an additional rule that prevents it from ever happening, eg. using the `updatedAt`-field.

**Comment by ganigeorgiev** [comment, marked as answer]
There is no built-in mechanism for resolving or preventing write conflicts and the last processed request will always "win".

The system `updated` field cannot be manually set via the web APIs, but if you send the previously fetched value with the request body, then you'll be able to compare it against the last `updated` record value in the database.

Update API rule:
```js
updated = @request.data.updated
```
_(aka, allow updates only if the previously fetched `updated` value matches with the last submitted one)_

Example scenario:
```js
// Bob fetch the ABC record
let record = await pb.collection("example").getOne("ABC");

// ... in the meantime Alice has already changed the ABC record ...

// Bob tries to update the ABC record
await pb.collection("example").update("ABC", {
    "updated": record.updated,
    "title": "test",
    // ....
})

// the update request will fail

```

If you want to be able to manually store the last update date, you can also create your own `date` type field, but similar to the above it is up to the developer to fill and check its value.

**Reply by einarpersson** [reply]
yes, this was my idea actually! glad to hear that it will work

**Reply by einarpersson** [reply]
Trying this out now. As far as I can tell, there is no way of identifying what rule that has been violated (so I cannot communicate what has happened to the user). Is that correct?

```
 response: {
    code: 404,
    message: "The requested resource wasn't found.",
    data: {}
  }
```

**Reply by ganigeorgiev** [reply]
@einarpersson Yes, that's correct because there is no easy way to detect which exact part of the rule caused the query to fail.

If you want more advanced error response, you can attach to the [`OnRecordBefore*Request` hooks](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks) and perform the data validation from there.

**Reply by einarpersson** [reply]
Is this an example of something that will be possible to do in JavaScript (without any rebuilding of the binary etc) once the [JavaScript events hooks](https://github.com/orgs/pocketbase/projects/2?pane=issue&itemId=17629907) or is it unrelated?

**Reply by ganigeorgiev** [reply]
@einarpersson Yes, eventually there will be support for binding to the same app event hooks with JS.

For this particular use case, I personally wouldn't bother with custom errors, because the create/update operations could generally have 2 error outcomes:
- the above generic error as a result of failed API rule
- field validation errors like non-empty, unique, min-length, etc. (in this case the `data` prop will be filled with the individual error field(s)).

Or in other words, you can "distinguish" to some extend the stale data errors from the regular validation errors by checking if the `data` prop is empty and show the user a generic error message suggesting that someone else has edited the record and the local data needs to be updated.

**Reply by einarpersson** [reply]
Thank you once again @ganigeorgiev for prompt responses. I just went ahead and made a minor donation (i'll try to get my organization to do so as well)
